### ISSUE TITLE
Build error with Next.js 15 in monorepo

### ISSUE THREAD

**Issue by omarshehab221** [vercel/next.js#72899]
### Link to the code that reproduces this issue

https://github.com/omarshehab221/DoomUI

### To Reproduce

Just clone the repo and try to build the apps/template or the apps/playground (I made sure not change anything in this one from what `bun create next-app@latest` created to test the error)

### Current vs. Expected behavior

I was trying to build the apps/template without that much change from what `bun create next-app@latest` created. I expected it to build fine specially that it passed the optimized build and the linting phases. It throws this error in the Collecting pages data phase: "Error: Minified React error #31; visit https://reactjs.org/docs/error-decoder.html?invariant=31&args[]=object%20with%20keys%20%7B%24%24typeof%2C%20type%2C%20key%2C%20ref%2C%20props%7D for the full message or use the non-minified dev environment for full errors and additional helpful warnings."

I created apps/playground and made sure not to change any of its code just to be sure the issue exists and had the same problem.

**This usually doesn't happen when I'm not working in a monorepo**

### Provide environment information

```bash
Operating System:
  Platform: win32
  Arch: x64
  Version: Windows 10 Pro
  Available memory (MB): 16272
  Available CPU cores: 4
Binaries:
  Node: 18.18.0
  npm: 10.5.0
  Yarn: 1.22.22
  pnpm: 9.9.0
  bun: 1.1.33
Relevant Packages:
  next: 15.0.3
  eslint-config-next: 15.0.3
  react: 18.3.1
  react-dom: 18.3.1
  typescript: 5.6.3
Next.js Config:
  output: N/A
```

### Which area(s) are affected? (Select all that apply)

create-next-app, Output (export/standalone), Turbopack

### Which stage(s) are affected? (Select all that apply)

next build (local)

### Additional context

I created apps/playground and didn't change it's code just to test this

**Comment by body20002** [comment]
A workaround for this issue is to not use turbopack,
this bug was open in [turborepo/#6298](https://github.com/vercel/turborepo/issues/6298) before moving turbopack to `vercel/nextjs` repo.

**Comment by sebnun** [comment]
I'm also having this issue with nextjs as a package in a monorepo (npm workspaces)
Note that nextjs works as expected with ```npm run dev``` which runs ````next dev --turbopack```, the error is shown when running ```npm run build```

I also tested running ```npx create-next-app@latest``` answering 'no' to ```Would you like to use Turbopack for next dev?```, but the error is still there when building.

**Comment by gameriff** [comment]
> A workaround for this issue is to not use turbopack, this bug was open in [turborepo/#6298](https://github.com/vercel/turborepo/issues/6298) before moving turbopack to `vercel/nextjs` repo.

The issue you reference seems to be in regard to using turbopack for dev, but this issue (which I am also experiencing) is referring to running a full build in a monorepo, if I am not mistaken.

**Comment by body20002** [comment]
@gameriff yeah sorry I didn't try to build before commenting and I was facing issues with a monorepo setup with nextjs 15, building seems to be fine on my end but again I'm not using turborepo just bun workspaces.

**Comment by Thaigun** [comment]
I can confirm the same issue, trying to build a project created by create-next-app. I tried with both Bun and Node and did not select Turbopack when setting up. I'm also in a workspaces monorepo.

**Comment by gameriff** [comment]
I was able to get my workspaces monorepo (note: I manually created this monorepo, i.e., not using create-react-app or other init script) to build with Next 15.0.3 and react/react-dom 19 release candidate versions by:

- Ensuring that my root package.json has the recommended types-react overrides set, as detailed here [React 19 RC Upgrade Guide](https://react.dev/blog/2024/04/25/react-19-upgrade-guide):

`{
  "dependencies": {
    "@types/react": "npm:types-react@rc",
    "@types/react-dom": "npm:types-react-dom@rc"
  },
  "overrides": {
    "@types/react": "npm:types-react@rc",
    "@types/react-dom": "npm:types-react-dom@rc"
  }
}`

and by removing any @types/react dependencies in any of the workspace package.jsons. 

And then:

- Installing the same react 19 release candidate version (whichever you choose, check NPM for latest) in all workspace packages, which, in my case, with my other dependencies, required using the `--force --legacy-peer-deps` flags to get a successful installation.

- I did end up deleting `.next/cache`, my `package-lock.json` and all of my `node_modules` directories at some point as well, as I was experimenting, but hard to tell if this was necessary or not.

**Comment by markomitranic** [comment]
This issue made me go insane until i **carefully** read the comment https://github.com/vercel/next.js/issues/72899#issuecomment-2501695993 from @gameriff .

I was already doing all those things, but it didn't work. The trick was to use **BOTH** `--force` and `--legacy-peer-deps` flags. Thank you <3 

I hate this.

**Comment by JeffBaumgardt** [comment]
Has anyone had luck with the above but where you have multiple apps with different versions of React/Dom? I have some projects that have dependencies that have yet to be updated to React@19 and mutiple attempts to brute force the update has failed. So for these apps I'm stuck on react@18 for the short term.

But when I start a new app I would want to start as fresh as prod ready versions go. This would of course be all sorts of dependencies that I am not worried about their installation target being mono root or app root. My CI should perform all install steps as needed.

**Comment by drakh** [comment]
Yeah, so i have a monorepo with strapi and nextJS..

Using pure npm workspaces i got to the situation in which either NextJS build works or strapi build works. never both...

So i ended up using lerna legacy package management. That way, everything works just fine.

**Comment by Jaimayal** [comment]
Ran into the same issue, building using `--turbopack` fixed the issue for me. Just dont forget to remove source maps if you're going to deploy to production

**Comment by timneutkens** [comment]
The most common case of failing to import monorepo code is that the project root can't be detected: https://nextjs.org/docs/app/api-reference/config/next-config-js/turbopack#root-directory.

Since the reproduction is a bit outdated I can't verify if it's something else, but there hasn't been new actionable reports here with a new reproduction so I'll close the issue. If you're running into something like this please open a new issue with a reproduction and we'd be happy to take a look.
